store/driver/txn/error.go

```diff
@@ -0,0 +1,225 @@
+			}
+		}
+		return genKeyExistsError(name, handle.String(), nil)
+	}
+
+	if len(value) == 0 {
+		return genKeyExistsError(name, handle.String(), errors.New("missing value"))
+	}
+
+	idxInfo := tables.FindPrimaryIndex(tblInfo)
+	if idxInfo == nil {
+		return genKeyExistsError(name, handle.String(), errors.New("cannot find index info"))
+	}
+
+	cols := make(map[int64]*types.FieldType, len(tblInfo.Columns))
+	for _, col := range tblInfo.Columns {
+		cols[col.ID] = &col.FieldType
+	}
+	handleColIDs := make([]int64, 0, len(idxInfo.Columns))
+	for _, col := range idxInfo.Columns {
+		handleColIDs = append(handleColIDs, tblInfo.Columns[col.Offset].ID)
+	}
+
+	row, err := tablecodec.DecodeRowToDatumMap(value, cols, time.Local)
+	if err != nil {
+		return genKeyExistsError(name, handle.String(), err)
+	}
+
+	data, err := tablecodec.DecodeHandleToDatumMap(handle, handleColIDs, cols, time.Local, row)
+	if err != nil {
+		return genKeyExistsError(name, handle.String(), err)
+	}
+
+	valueStr := make([]string, 0, len(data))
+	for _, col := range idxInfo.Columns {
+		d := data[tblInfo.Columns[col.Offset].ID]
+		str, err := d.ToString()
+		if err != nil {
+			return genKeyExistsError(name, key.String(), err)
+		}
+		valueStr = append(valueStr, str)
+	}
+	return genKeyExistsError(name, strings.Join(valueStr, "-"), nil)
+}
+
+func extractKeyExistsErrFromIndex(key kv.Key, value []byte, tblInfo *model.TableInfo, indexID int64) error {
+	var idxInfo *model.IndexInfo
+	for _, index := range tblInfo.Indices {
+		if index.ID == indexID {
+			idxInfo = index
+		}
+	}
+	if idxInfo == nil {
+		return genKeyExistsError("UNKNOWN", key.String(), errors.New("cannot find index info"))
+	}
+	name := idxInfo.Name.String()
+
+	if len(value) == 0 {
+		return genKeyExistsError(name, key.String(), errors.New("missing value"))
+	}
+
+	colInfo := tables.BuildRowcodecColInfoForIndexColumns(idxInfo, tblInfo)
+	values, err := tablecodec.DecodeIndexKV(key, value, len(idxInfo.Columns), tablecodec.HandleNotNeeded, colInfo)
+	if err != nil {
+		return genKeyExistsError(name, key.String(), err)
+	}
+	valueStr := make([]string, 0, len(values))
+	for i, val := range values {
+		d, err := tablecodec.DecodeColumnValue(val, colInfo[i].Ft, time.Local)
+		if err != nil {
+			return genKeyExistsError(name, key.String(), err)
+		}
+		str, err := d.ToString()
+		if err != nil {
+			return genKeyExistsError(name, key.String(), err)
+		}
+		valueStr = append(valueStr, str)
+	}
+	return genKeyExistsError(name, strings.Join(valueStr, "-"), nil)
+}
+
+func extractKeyErr(err error) error {
+	if err == nil {
+		return nil
+	}
+	if e, ok := errors.Cause(err).(*tikvstore.ErrWriteConflict); ok {
+		return newWriteConflictError(e.WriteConflict)
+	}
+	if e, ok := errors.Cause(err).(*tikvstore.ErrRetryable); ok {
+		notFoundDetail := prettyLockNotFoundKey(e.Retryable)
+		return kv.ErrTxnRetryable.GenWithStackByArgs(e.Retryable + " " + notFoundDetail)
+	}
+	return errors.Trace(err)
```

Comment:
As there could be more and more types of errors, what about using `switch errors.Cause(err).(type)`?